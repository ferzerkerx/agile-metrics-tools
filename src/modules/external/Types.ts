export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  srcPath: string;
}

export class ExternalData {
  createdAt: Date;
  [type: string]: any;
}

export interface ExternalService {
  fetchExternalInfo(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
