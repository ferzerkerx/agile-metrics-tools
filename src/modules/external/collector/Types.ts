import { CollectorConfig, MetricItem } from "../../../metrics";

export interface ExternalMetricItem extends MetricItem {}

export class ExternalCollectorConfig implements CollectorConfig {
  since: Date;
  type: string;
  srcType: string;
  srcPath: string;
  metricType: string;
  constructor({ since, type, srcType, srcPath, metricType }) {
    this.since = since;
    this.type = type;
    this.srcType = srcType;
    this.srcPath = srcPath;
    this.metricType = metricType;
  }
}
