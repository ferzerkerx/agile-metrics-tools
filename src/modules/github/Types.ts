export class GithubCommit {
  sha: string;
  createdAt: Date;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  message: string;
  pullRequest?: PullRequest;
}

export class PullRequest {
  id: number;
  numberOfDaysOpen: number;
  numberOfComments: number;
}

export interface GithubConfig {
  repositoryName: string;
  orgName: string;
  since: string;
}

export interface GithubService {
  commits(githubConfig: GithubConfig): Promise<GithubCommit[]>;
}

export interface GithubClient {
  commits(githubConfig: GithubConfig): Promise<string[]>;

  getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<any>;

  pullRequestForCommit(sha: string): Promise<any>;
}
