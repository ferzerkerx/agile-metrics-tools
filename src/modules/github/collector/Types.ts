import { CollectorConfig, MetricItem } from '../../../metrics';

export interface GithubMetricItem extends MetricItem {
  sha: string;
  linesAdded: number;
  linesRemoved: number;
  author: string;
  repositoryName: string;
  pullRequest?: {
    prId: number;
    numberOfDaysOpen: number;
    numberOfComments: number;
  };
}

export class GithubCollectorConfig implements CollectorConfig {
  repositoryName: string;
  orgName: string;
  since: string;

  constructor({ repositoryName, orgName, since }) {
    this.repositoryName = repositoryName;
    this.orgName = orgName;
    this.since = since;
  }
}
