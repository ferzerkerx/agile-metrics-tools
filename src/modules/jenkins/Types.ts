export class JenkinsJob {
  name: string;
  branch: string;
  builds: JenkinsBuild[];
}

export class JenkinsBuild {
  timestamp: Date;
  result: string;
  durationInMs: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  url: string;
}

export interface JenkinsService {
  findData(
    orgName: string,
    projectName: string,
    since: Date
  ): Promise<JenkinsJob>;
}

export interface JenkinsClient {
  getData(url): Promise<any>;
}
