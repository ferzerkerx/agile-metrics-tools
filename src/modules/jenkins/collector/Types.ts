import { CollectorConfig, MetricItem } from "../../../metrics/Types";

export interface JenkinsMetricItem extends MetricItem {
  jenkinsUrl: string;
  buildName: string;
  result: string;
  durationInSeconds: number;
  causedBy: string;
  revision: string;
  revisionDescription: string;
  buildNumber: string;
  projectName: string;
}

export class JenkinsCollectorConfig implements CollectorConfig {
  orgName: string;
  projectName: string;
  since: Date;

  constructor({ orgName, projectName, since }) {
    this.orgName = orgName;
    this.projectName = projectName;
    this.since = new Date(since);
  }
}
