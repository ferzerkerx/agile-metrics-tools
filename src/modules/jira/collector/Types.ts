import { CollectorConfig, MetricItem } from '../../../metrics';

export interface JiraMetricItem extends MetricItem {
  key: string;
  createdBy: string;
  issueType: string;
  movedForward: number;
  movedBackward: number;
  storyPoints: number;
  assignees: string[];
  tags: string[];
  created: Date;
  finished: Date;
  leadTime: number;
  devTime: number;
  commentCount: number;
  jiraProject: string;
  teamName: string;
  estimateHealth: number;
  rawEstimateHealth: number;
  numberOfBugs: number;
}

export class JiraCollectorConfig implements CollectorConfig {
  teamId: number;
  teamName: string;
  workFlowType: 'sprint' | 'kanban';
  since: Date;
  until?: Date;
  workFlowMap?: { [name: string]: number };
  fields?: { [name: string]: string };
  estimateConfig?: { maxTime: number; estimationValues: number[] };

  constructor({
    teamId,
    since,
    until = null,
    workFlowMap,
    fields,
    teamName,
    workFlowType
  }) {
    this.teamId = teamId;
    this.since = new Date(since);
    this.until = until ? new Date(until) : null;
    this.workFlowMap = workFlowMap;
    this.fields = fields;
    this.teamName = teamName;
    this.workFlowType = workFlowType ? workFlowType : 'sprint';
  }

  isKanban() {
    return this.workFlowType === 'kanban';
  }

  isSprint() {
    return this.workFlowType === 'sprint';
  }
}
