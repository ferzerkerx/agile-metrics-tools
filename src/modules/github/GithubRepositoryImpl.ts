import {GithubClient, GithubCommit, GithubConfig, GithubRepository} from "./Types";
import { Converters } from "./Converters";

export class GithubRepositoryImpl implements GithubRepository {
  constructor(private githubClient: GithubClient) {}

  async commits(githubConfig: GithubConfig): Promise<GithubCommit[]> {
    const { repositoryName, orgName } = githubConfig;

    const commitsSha: string[] = await this.githubClient.commits(githubConfig);
    const commitsWithDetailsPromise: Promise<GithubCommit>[] = commitsSha.map(
      commitSha => {
        return this.getCommitDetails(repositoryName, orgName, commitSha);
      }
    );
    return await Promise.all(commitsWithDetailsPromise);
  }

  private async getCommitDetails(
    repositoryName: string,
    orgName: string,
    sha: string
  ): Promise<GithubCommit> {
    const commitDetailsResponse = await this.githubClient.getCommitDetails(
      repositoryName,
      orgName,
      sha
    );

    const commitPrResponse = await this.githubClient.pullRequestForCommit(sha);

    return Converters.toGithubCommit(commitDetailsResponse, commitPrResponse);
  }
}
