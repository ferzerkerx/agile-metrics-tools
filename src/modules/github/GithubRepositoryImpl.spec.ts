import { GithubClient, GithubConfig, GithubRepository } from './Types';
import * as commitsFakeData from './test/commits-response.json';
import * as commitDetailsFakeData from './test/commit-details-response.json';
import * as commitPrFakeData from './test/commit-pr-response.json';
import * as commitPrReviewComentsFakeData from './test/pull-request-review-comments-response.json';
import { GithubRepositoryImpl } from './GithubRepositoryImpl';

describe('GithubRepository', () => {
  const githubClient: GithubClient = {
    async commits(githubConfig: GithubConfig): Promise<any> {
      return commitsFakeData;
    },
    async getCommitDetails(
      repositoryName: string,
      orgName: string,
      sha: string
    ): Promise<any> {
      return commitDetailsFakeData;
    },
    async pullRequestForCommit(sha: string): Promise<any> {
      return commitPrFakeData;
    },
    async pullRequestComments({
      owner,
      repo,

      number
    }): Promise<any> {
      return commitPrReviewComentsFakeData;
    }
  };

  const githubService: GithubRepository = new GithubRepositoryImpl(
    githubClient
  );

  it('should get commits', async () => {
    const githubConfig: GithubConfig = {
      repositoryName: 'someRepositoryName',
      orgName: 'someOrgName',
      since: '2018-12-03'
    };
    const data = await githubService.commits(githubConfig);
    expect(data).toMatchSnapshot();
  });
});
