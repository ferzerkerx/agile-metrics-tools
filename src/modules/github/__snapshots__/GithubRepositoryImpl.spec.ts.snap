// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubRepository should get commits 1`] = `
Array [
  Object {
    "author": "somUser@org",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "linesAdded": 0,
    "linesRemoved": 0,
    "message": "ORG-5917
- some comment",
    "pullRequest": Object {
      "numberOfComments": 6,
      "numberOfDaysOpen": 0,
      "prId": 177,
    },
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "somUser@org",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "linesAdded": 0,
    "linesRemoved": 0,
    "message": "ORG-5917
- some comment",
    "pullRequest": Object {
      "numberOfComments": 6,
      "numberOfDaysOpen": 0,
      "prId": 177,
    },
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "somUser@org",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "linesAdded": 0,
    "linesRemoved": 0,
    "message": "ORG-5917
- some comment",
    "pullRequest": Object {
      "numberOfComments": 6,
      "numberOfDaysOpen": 0,
      "prId": 177,
    },
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "somUser@org",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "linesAdded": 0,
    "linesRemoved": 0,
    "message": "ORG-5917
- some comment",
    "pullRequest": Object {
      "numberOfComments": 6,
      "numberOfDaysOpen": 0,
      "prId": 177,
    },
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
  Object {
    "author": "somUser@org",
    "createdAt": 2018-12-13T14:33:21.000Z,
    "linesAdded": 0,
    "linesRemoved": 0,
    "message": "ORG-5917
- some comment",
    "pullRequest": Object {
      "numberOfComments": 6,
      "numberOfDaysOpen": 0,
      "prId": 177,
    },
    "sha": "2a20a913d0159b1105e1be9f398d95a73e1a8129",
  },
]
`;
